homeassistant/core.py

```diff
@@ -281,14 +283,35 @@ def async_add_job(
         return task
 
     @callback
-    def async_create_task(self, target: Coroutine) -> asyncio.tasks.Task:
+    def async_catching_create_task(
+            self, target: Coroutine) -> asyncio.tasks.Task:
+        """Create a task from within the eventloop and print exception.
+
+        This method must be run in the event loop.
+
+        target: target to call.
+        """
+        self.async_create_task(target, True)
+
+    @callback
+    def async_create_task(
+            self, target: Coroutine, catch=False) -> asyncio.tasks.Task:
         """Create a task from within the eventloop.
 
         This method must be run in the event loop.
 
         target: target to call.
         """
-        task = self.loop.create_task(target)  # type: asyncio.tasks.Task
+        trace = traceback.extract_stack()
+        wrapped_target = target
+        if catch:
+            wrapped_target = catch_log_exception(
```

Comment:
Instead of doing this inside `core.py`, I think that we should do it at the places that call `async_create_task`. That way we can keep the API surface of the core minimal.

For example, now the stack is extracted on each async_create_task, even when not used. Core functionality should be fast and not do extra work.

```python
hass.helpers.logging.async_create_catching_task(my_coro())
```

Reply:
> For example, now the stack is extracted on each async_create_task, even when not used.

Oops, that was a mistake and not intentional! Extracting of the stack should of course be moved:
```diff
-        trace = traceback.extract_stack()
         wrapped_target = target
         if catch:
+            trace = traceback.extract_stack()
             wrapped_target = catch_log_exception(
```

Anyhow, I agree it's reasonably clean to for example replace:
```python
self.hass.async_create_task(self.async_set_cast_info(discover))
```
With:
```python
discover = logging.async_create_catching_task(self.async_set_cast_info(discover))
self.hass.async_create_task(discover)
```

But it becomes a bit complex, we need to define also a `logging.async_create_catching_callback` as well as some auto-detecting variant to support `async_run_job` and `async_add_job`, and the correct wrapper must then be paired with the correct core function.

Also, `async_create_task` currently occurs 280 times in 149 files in the source tree.
259 of the occurrences are in the `components` subtree, which I think could motivate an "opt-out" where core functions skip the overhead.

Reply:
So we need to consider what we're trying to solve: proper attribution of when something breaks.

Historically, we have some mistakes around this. We would prefer to do create_task instead of calling functions directly, this was the right approach before we used asyncio, but it no longer is.

So any await context that uses async_create_task should probably await it, especially if it is the last line of the method.

The main reason we use async_add_job, is because we don't know what kind of function we got in. However, most of the time we do, and we can/should resolve that in our event helpers. All cases that we know what type of method, we should use the appropriate method (create_task, create_executor_job). And again, await as much as possible to keep track of context.

I am a firm believer of doing things when it's needed, and only then. So instead of putting the logic inside `async_create_task`, I rather have the logging task be created outside of it, and then passed a logging coroutine to async_create_task. Given that both `async_add_executor_job` and `async_create_task` return an awaitable object (a task), it will be easy to wrap.